chronos/apps/http/httpclient.nim

```diff
@@ -604,18 +666,35 @@ proc acquireConnection(
   else:
     let conn =
       block:
-        let conns = session.connections.getOrDefault(ha.id)
-        if len(conns) > 0:
-          var res: HttpClientConnectionRef = nil
-          for item in conns:
-            if item.state == HttpClientConnectionState.Ready:
-              res = item
-              break
-          res
-        else:
-          nil
+        var idleConnections: seq[HttpClientConnectionRef]
+        let
+          timestamp = Moment.now()
+          res =
+            block:
+              var cres: HttpClientConnectionRef = nil
+              session.connections.withValue(ha.id, connections):
+                for connection in connections[]:
+                  if connection.isReady():
+                    if connection.isIdle(timestamp, session.idleTimeout):
+                      idleConnections.add(connection)
+                    else:
+                      if isNil(cres): cres = connection
+              cres
+
+        if not(isNil(res)):
+          # We mark connection as acquired to avoid race, because we need to
+          # prune idle connections.
+          res.state = HttpClientConnectionState.Acquired
+        if len(idleConnections) > 0:
+          # Closing connections which was idle for`session.idleTimeout`
+          # time.
+          var pending: seq[Future[void]]
+          for connection in idleConnections:
+            pending.add(session.removeConnection(connection))
+          await allFutures(pending)
```

Comment:
this looks risky: if I'm trying to acquire a connection presumably I have some use for it which shouldn't be delayed by the closing of unrelated connections